### ISSUE TITLE
propose 'no-reassign-property'

### ISSUE THREAD

**Issue by ZYinMD** [eslint/eslint#12355]
**Please describe what the rule should do:**
Warn if you assign an object's property to a variable, then re-assign that variable, because you may mistakenly think the object's property will change too.

**What category of rule is this? (place an "X" next to just one item)**

[x] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
let myFlag = window.foo.bar.baz.myFlag;
myFlag = false; // danger: your may have intended to change window.foo.bar.baz.myFlag 

let { myArray } = data;
myArray = myArray.filter(xxx); // danger: your may have intended to change data.myArray 
```

**Why should this rule be included in ESLint (instead of a plugin)?**
It covers the same kind of danger as `no-param-reassign` does. 
In `no-param-reassign`, you may mistakenly think you can change the value of a param just like you can mutate a param, but your change will have no effect.
When this rule doesn't cover a certain case, `no-param-reassign` probably does.
When `no-param-reassign` doesn't cover a certain case, this rule probably does.

**Are you willing to submit a pull request to implement this rule?**
I'm not capable.

**Comment by ljharb** [comment]
This is a fundamental piece of the language; it would warn on the vast majority of the code i've ever written. I don't think it's appropriate for a lint rule to warn on something that you have to already know to be able to write JavaScript in the first place.

**Comment by kaicataldo** [comment]
Thanks for the proposal. I agree with @ljharb. The examples given are valid code and this feels like a rule that is overreaching in what it warns to me. If this is something you would like to enforce in your own codebases, I think this is a better fit for a custom rule.
